programs/ssl/ssl_client2.c

```diff
@@ -3102,10 +3117,56 @@ int main( int argc, char *argv[] )
 
         mbedtls_printf( " ok\n" );
 
+        /* Save serialized context to the 'opt.serial_to_file' as a base64 code */
+        if( 0 < strlen( opt.serial_to_file ) )
+        {
+            FILE *b64_file;
+            uint8_t *b64_buf;
+            size_t b64_len;
+
+            mbedtls_printf( "  . Save serialized context to a file... " );
+
+            mbedtls_base64_encode( NULL, 0, &b64_len, context_buf, buf_len );
+
+            if( ( b64_buf = mbedtls_calloc( 1, b64_len ) ) == NULL )
+            {
+                mbedtls_printf( "failed\n  ! Couldn't allocate buffer for "
+                                "the base64 code\n" );
+                goto exit;
+            }
+
+            if( ( ret = mbedtls_base64_encode( b64_buf, b64_len, &b64_len,
+                                               context_buf, buf_len ) ) != 0 )
+            {
+                mbedtls_printf( "failed\n  ! mbedtls_base64_encode returned "
+                            "-0x%x\n", -ret );
+                goto exit;
+            }
+
+            b64_file = fopen( opt.serial_to_file, "w" );
+            if( NULL == b64_buf )
+            {
+                mbedtls_printf( "failed\n  ! Cannot open '%s' for writing.\n",
+                                opt.serial_to_file );
+                goto exit;
+            }
+
+            if( b64_len != fwrite( b64_buf, 1, b64_len, b64_file ) )
+            {
+                mbedtls_printf( "failed\n  ! fwrite(%ld bytes) failed\n",
+                                (long) b64_len );
+                fclose( b64_file );
+                goto exit;
+            }
+
+            mbedtls_printf( "ok\n" );
+            fclose( b64_file );
+        }
+
         if( opt.serialize == 1 )
         {
             /* nothing to do here, done by context_save() already */
-            mbedtls_printf( "  . Context has been reset... ok" );
+            mbedtls_printf( "  . Context has been reset... ok\n" );
```

Comment:
This looks like an unrelated change. Is it needed? If so, could we do it in its own commit?

Reply:
Rather needed. Without it the log looked like this:
```
  . Serializing live connection... ok
  . Context has been reset... ok  . Deserializing connection... ok
  . Closing the connection... done
```
I put it to separate commit

Reply:
Added missing mbedtls_free to fix CI failure

Reply:
Thanks